src/main/java/part2/cache/CachingDataStorageImpl.java

```diff
@@ -38,6 +39,25 @@ public CachingDataStorageImpl(DataStorage<String, T> db, int timeout, TimeUnit t
         // TODO don't use obtrudeException on result - just don't
         // TODO use remove(Object key, Object value) to remove target value
         // TODO Start timeout after receiving result in CompletableFuture, not after receiving CompletableFuture itself
-        throw new UnsupportedOperationException();
+        final OutdatableResult<T> oldResult = cache.get(key);
+        if (oldResult != null) {
+            return oldResult;
+        }
+        final CompletableFuture<Void> outdatedCF = new CompletableFuture<>();
+        final CompletableFuture<T> resultCF = new CompletableFuture<>();
+        final OutdatableResult<T> newResult = new OutdatableResult<>(resultCF, outdatedCF);
+        cache.put(key, newResult);

```

Comment:
You can override an existing value. Note that `cache.get() + cache.put()` is not an atomic operation. Use `putIfAbsent`